lib/facade_converter.ts

```diff
@@ -354,6 +377,73 @@ export class FacadeConverter extends base.TranspilerBase {
     },
   };
 
+  private es6Promises: ts.Map<CallHandler> = {
+    'Promise.resolve': (c: ts.CallExpression, context: ts.Expression) => {
+      this.emit('Future.value(');
+      this.visitList(c.arguments);  // Should only be 1.
+      this.emit(')');
+    },
+    'Promise.reject': (c: ts.CallExpression, context: ts.Expression) => {
+      this.emit('Future.error(');
+      this.visitList(c.arguments);  // Should only be 1.
+      this.emit(')');
+    },
+    'Promise.catch': (c: ts.CallExpression, context: ts.Expression) => {
+      this.visit(context);
+      this.emit('.catchError(');
+      this.visitList(c.arguments);
+      this.emit(')');
+    },
+    'Promise': (c: ts.CallExpression, context: ts.Expression) => {
+      if (c.kind != ts.SyntaxKind.NewExpression) {
+        return true;
+      }
+      this.assert(c, c.arguments.length == 1, 'Promise construction must take 2 arguments.');
+      this.assert(
+          c, c.arguments[0].kind == ts.SyntaxKind.ArrowFunction ||
+              c.arguments[0].kind == ts.SyntaxKind.FunctionExpression,
+          'Promise argument must be a function expression (or arrow function).');
+      let callback: ts.FunctionLikeDeclaration;
+      if (c.arguments[0].kind == ts.SyntaxKind.ArrowFunction)
+        callback = <ts.FunctionLikeDeclaration>(<ts.ArrowFunction>c.arguments[0]);
+      if (c.arguments[0].kind == ts.SyntaxKind.FunctionExpression)
+        callback = <ts.FunctionLikeDeclaration>(<ts.FunctionExpression>c.arguments[0]);
+      this.assert(
+          c, callback.parameters.length > 0 && callback.parameters.length < 3,
+          'Promise executor must take 1 or 2 arguments (resolve and reject).');
+
+      this.assert(
+          c, callback.parameters[0].name.kind == ts.SyntaxKind.Identifier,
+          'First argument of the Promise executor is not a straight parameter.');
+      let resolveParameterIdent = <ts.Identifier>(callback.parameters[0].name);
+      this.assert(
+          c, resolveParameterIdent.text != '_completer',
+          'Promise executor arguments cannot have "_completer" as a name.');
+
+      this.emit('(() {');  // Create a new scope.
```

Comment:
Actually I cannot get rid of that scope for various reason:

1. The point where I'm at in the emit stream is `Future someVarName = >>>here<<<`. Which means that I can't rewrite the declaration.
2. Having access to the `Future` variable name would be really hard as well, because multiple places in the ancestry could be defining it. It could be a key on an object... :/ Or it could be entirely unnamed.
3. I've re-added the inner function scope because the executor might return, and in this case that would break the generated code :/

If you find solutions for some of these I can get it in a better shape. Unless we actually do AST => AST transform at some point this will be a really hard problem.